feat:

KlochkovDimka opened the pull request:
1. Добавление функций в классе UserController.java: добавления в друзья, удаление из друзей, получение пользователя по идентификационному номеру, получение списка друзей, получение списка общих друзей;
2. Добавление класса UserService.java для работы контроллера UserController;
3. Добавления класса InMemoryUserStorage.java для хранения и обработки пользователей;
4. Добавление функций в классе FilmController.java: добавление лайков пользователями, удаление лайков пользователями, получение списка популярных фильмов;
5. Добавление класса FilmService.java для работы контроллера FilmController;
6. Добавление класса InMemoryFilmStorage.java для хранения и обработки фильмов;
7. Добавление класса ErrorHandler для обработки исключений;
8. Добавление в классы: FilmControllerTest.java, UserControllerTest.java, тестов для новых функций.

KlochkovDimka committed:
feat:
1. Добавление функций в классе UserController.java: добавления в друзья, удаление из друзей, получение пользователя по идентификационному номеру, получение списка друзей, получение списка общих друзей;
2. Добавление класса UserService.java для работы контроллера UserController;
3. Добавления класса InMemoryUserStorage.java для хранения и обработки пользователей;
4. Добавление функций в классе FilmController.java: добавление лайков пользователями, удаление лайков пользователями, получение списка популярных фильмов;
5. Добавление класса FilmService.java для работы контроллера FilmController;
6. Добавление класса InMemoryFilmStorage.java для хранения и обработки фильмов;
7. Добавление класса ErrorHandler для обработки исключений;
8. Добавление в классы: FilmControllerTest.java, UserControllerTest.java, тестов для новых функций.

KlochkovDimka committed:
feat:
1. Добавление функций в классе UserController.java: добавления в друзья, удаление из друзей, получение пользователя по идентификационному номеру, получение списка друзей, получение списка общих друзей;
2. Добавление класса UserService.java для работы контроллера UserController;
3. Добавления класса InMemoryUserStorage.java для хранения и обработки пользователей;
4. Добавление функций в классе FilmController.java: добавление лайков пользователями, удаление лайков пользователями, получение списка популярных фильмов;
5. Добавление класса FilmService.java для работы контроллера FilmController;
6. Добавление класса InMemoryFilmStorage.java для хранения и обработки фильмов;
7. Добавление класса ErrorHandler для обработки исключений;
8. Добавление в классы: FilmControllerTest.java, UserControllerTest.java, тестов для новых функций.

LexLippi commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -14,46 +17,60 @@
 
-    private Map<Integer, Film> films = new HashMap();
-    private int generatedId = 1;
+    private final InMemoryFilmStorage inMemoryFilmStorage;
```
1. В терминах слоеной архитектуры, контроллер не должен знать ничего про хранилище, оно должно абстрагироваться с помощью сервиса
2. Не стоит использовать явный тип хранилища, его надо абстрагировать интерфейсом

LexLippi commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -14,46 +17,60 @@
+
+    @GetMapping("/popular")
+    public ArrayList<Film> getPopularFilm(
+            @RequestParam(defaultValue = "10", required = false) Integer count) {
```
Если указываем `defaultValue`, то нет необходимости в `required = false`

LexLippi commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -1,52 +1,83 @@
 @RestController
 @RequestMapping("/users")
 @Slf4j
 public class UserController {
```
Замечания аналогичные `FilmController`

LexLippi commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -4,13 +4,17 @@
 @Data
 public class Film {
 
+    private Set<Integer> likesFromUsers = new HashSet<>();
```
Для этого поля стоит переопределить метод `get`, так как возвращение самой коллекции влечет нарушения инкапсуляции и можно модифицировать данные с помощью этого метода, поэтому необходимо возвращать копию коллекции

LexLippi commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -4,10 +4,13 @@
 @Data
 public class User {
     private int id;
+    private Set<Integer> friends = new HashSet<>();
```
Замечание аналогичное `Film`

LexLippi commented on src/main/java/ru/yandex/practicum/filmorate/service/film/FilmService.java:
```diff
@@ -0,0 +1,58 @@
+@Slf4j
+public class FilmService {
+
+    private final InMemoryFilmStorage inMemoryFilmStorage;
+    private final InMemoryUserStorage inMemoryUserStorage;
```
Необходимо взаимодействовать с интерфейсами, а не с реализациями хранилищ

LexLippi commented on src/main/java/ru/yandex/practicum/filmorate/service/film/FilmService.java:
```diff
@@ -0,0 +1,58 @@
+        ArrayList<Film> listSortFilm = new ArrayList<>(inMemoryFilmStorage.getListFilms());
+
+        return listSortFilm.stream()
+                .sorted((f1, f2) -> {
```
Компаратор не меняется от вызова к вызову, стоит вынести его, как поле класса

LexLippi commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java:
```diff
@@ -0,0 +1,62 @@
+
+    //Проверка на наличие фильма по идентификационному номеру
+    public void isFilm(int filmId) {
+        if (!films.containsKey(filmId)) throw new NotExistFilmException("Фильма с данным id не существует");
```
В соответствии с `Java Google Code Style` советую ставить скобки даже если они являются опциональными:

https://google.github.io/styleguide/javaguide.html#s4.1.1-braces-always-used

LexLippi commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/InMemoryUserStorage.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    //Проверка на наличие пользователя
+    public void isUser(int idUser) {
```
Замечание аналогичное `FilmStorage`

LexLippi reviewed (changes requested):
Привет!

Неплохая работа!

Есть некоторое количество замечаний, которые надо исправить!

Желаю удачи!

KlochkovDimka committed:
feat:
Внесение поправок в классы: FilmController.java, UserController.java, Film.java, User.java, FilmService.java, UserService.java, FilmStorage.java, UserStorage.java, InMemoryFilmStorage.java, InMemoryUserStorage.java,

KlochkovDimka committed:
feat:
Внесение поправок в классы: FilmController.java, UserController.java, Film.java, User.java, FilmService.java, UserService.java, FilmStorage.java, UserStorage.java, InMemoryFilmStorage.java, InMemoryUserStorage.java,

LexLippi reviewed (approved):
Привет!

Хорошая работа!

Желаю удачи в следующих спринтах!
